micrometer-core/src/main/java/io/micrometer/core/instrument/binder/grpc/ObservationGrpcServerInterceptor.java

```diff
@@ -0,0 +1,122 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.micrometer.core.instrument.binder.grpc;
+
+import io.grpc.*;
+import io.grpc.Metadata.Key;
+import io.grpc.MethodDescriptor.MethodType;
+import io.grpc.ServerCall.Listener;
+import io.micrometer.common.lang.Nullable;
+import io.micrometer.observation.Observation;
+import io.micrometer.observation.Observation.Scope;
+import io.micrometer.observation.ObservationRegistry;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * A gRPC server interceptor that works with {@link Observation}.
+ * <p>
+ * <b>Usage:</b>
+ * </p>
+ * <pre>
+ * Server server = ServerBuilder.forPort(8080)
+ *         .intercept(new ObservationGrpcServerInterceptor(meterRegistry))
+ *         .build();
+ * server.start()
+ * </pre> The instrumentation is based on the behavior of Spring Cloud Sleuth and Brave.
+ *
+ * @author Tadaya Tsuyukubo
+ * @since 1.10.0
+ */
+public class ObservationGrpcServerInterceptor implements ServerInterceptor {
+
+    private static final GrpcServerObservationConvention DEFAULT_CONVENTION = new DefaultGrpcServerObservationConvention();
+
+    private static final Map<String, Key<String>> KEY_CACHE = new ConcurrentHashMap<>();
+
+    private final ObservationRegistry registry;
+
+    @Nullable
+    private GrpcServerObservationConvention customConvention;
+
+    public ObservationGrpcServerInterceptor(ObservationRegistry registry) {
+        this.registry = registry;
+    }
+
+    @Override
+    public <ReqT, RespT> Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call, Metadata headers,
+            ServerCallHandler<ReqT, RespT> next) {
+        Supplier<GrpcServerObservationContext> contextSupplier = () -> {
+            GrpcServerObservationContext context = new GrpcServerObservationContext((carrier, keyName) -> {
+                Key<String> key = KEY_CACHE.computeIfAbsent(keyName,
+                        (k) -> Key.of(keyName, Metadata.ASCII_STRING_MARSHALLER));
+                return carrier.get(key);
+            });
+            context.setCarrier(headers);
+
+            MethodDescriptor<ReqT, RespT> methodDescriptor = call.getMethodDescriptor();
+            String serviceName = methodDescriptor.getServiceName();
+            String methodName = methodDescriptor.getBareMethodName();
+            MethodType methodType = methodDescriptor.getType();
+            String spanName = methodDescriptor.getFullMethodName();
```

Comment:
Can we actually try to hide the details such as metric name or span name? It's really a contextual name (I know that at the end of the day a span will be named like this but imagine if someone uses the contextual name only for logging and doesn't have tracing on the classpath).

Reply:
The contextual name is now hidden and used in convention class only.